spring-cloud-gcp-starters/spring-cloud-gcp-starter-core/src/main/java/org/springframework/cloud/gcp/core/GCPProperties.java

```diff
@@ -27,7 +28,7 @@
 
 	private String projectId;
 
-	private String jsonKeyFile;
+	private Resource credentialKeyFile;
```

Comment:
Ahhh the `ProjectIdProvider` was the missing piece here. We should auto-configure `ProjectIdProvider` based on `GcpProperties.projectId` or other possible sources, and then client code can use that. `PubSubMessageHandler` and `PubSubTemplate` would be configured with `ProjectIdProvider` instead of `String projectId`. Am I understanding this correctly?